BlueprintUI/Sources/Internal/ElementIdentifier.swift

```diff
@@ -1,43 +1,114 @@
+ 
+ 2) The key. This is an optional value provided by developers using Blueprint to further
+ disambiguate identical elements within a hierarchy. This is optional, and is rarely provided.
+ When it is provided, `elementType` + ` key` is used to disambiguate elements.
+ 
+ 3) The occurrence count of that type of `elementType` + `key` in the hierarchy.
+ For example, if I have a hierarchy of [A, B, B] the counts will be [1, 1, 2] respectively.
 
-    /// The element represented by this component was not assigned a specific key, so its index in the parent's array
-    /// of children is used instead.
-    case index(Int)
+ A fully constructed `ElementIdentifier` would look like this:
+ 
+ ```
+ ElementIdentifier(elementType: MyElement.self, key: nil, count: 2)
+ ```
+ 
+ Which means that this identifier represents an element of type `MyElement`, which is
+ the second `MyElement` element in the hierarchy.
+ 
+ Elements are identified and counted by their `elementType` and `key` to further assist in diffing during update operations.
+ 
+ For example, if we started with:
+ ```
+ ElementA -> ElementIdentifier(ElementA.self, [optional key], 1)
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 2)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ And then removed A and the first B:
+ 
+ ```
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ You will note that the identifiers remain stable, which ultimately ensures that views are reused.
+ */
+struct ElementIdentifier: Hashable, CustomDebugStringConvertible {
+    
+    let elementType : ObjectIdentifier
+    let key : AnyHashable?
 
-    internal init(key: String?, index: Int) {
-        if let key = key {
-            self = .key(key)
+    let count : Int
+
+    init(elementType : Element.Type, key : AnyHashable?, count : Int) {
+        
+        self.elementType = ObjectIdentifier(elementType)
+        self.key = key
+        
+        self.count = count
+    }
+    
+    var debugDescription: String {
+        if let key = self.key {
+            return "\(self.elementType).\(String(describing: key)).\(self.count)"
         } else {
-            self = .index(index)
+            return "\(self.elementType).\(self.count)"
         }
     }
-
-    /// Returns the reuse identifier of this component (if provided).
-    var key: String? {
-        switch self {
-        case .key(let key):
-            return key
-        default:
-            return nil
+    
+    /**
+     Internal type used to create `ElementIdentifier` instances during view hierarchy updates.
+     */
+    struct Factory {
+        
+        init(reserveCapacity capacity: Int? = nil) {
+            
+            self.countsByKey = [:]
+            
+            if let capacity = capacity {
```

Comment:
Dictionary has an init that takes in a minimum capacity, this init could be simplified to just `self.countsByKey = .init(minimumCapacity: capacity ?? 0)` (and you could also make `capacity` non-optional with a default of 0 to avoid the `??`)

Reply:
Pretty much just style though, feel free to leave ¯\\\_(ツ)_/¯

Reply:
Didn't know about `Dictionary(minimumCapacity: capacity)`; switching over!